OsmAnd/src/net/osmand/plus/quickaction/actions/ShowHideGpxTracksAction.java

```diff
@@ -0,0 +1,86 @@
+package net.osmand.plus.quickaction.actions;
+
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.TextView;
+import android.widget.Toast;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import net.osmand.CallbackWithObject;
+import net.osmand.GPXUtilities.GPXFile;
+import net.osmand.IndexConstants;
+import net.osmand.plus.GpxSelectionHelper;
+import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.OsmandSettings;
+import net.osmand.plus.R;
+import net.osmand.plus.activities.MapActivity;
+import net.osmand.plus.helpers.GpxUiHelper;
+import net.osmand.plus.quickaction.QuickAction;
+
+public class ShowHideGpxTracksAction extends QuickAction {
+
+	public static final int TYPE = 28;
+
+	public ShowHideGpxTracksAction() {
+		super(TYPE);
+	}
+
+	public ShowHideGpxTracksAction(QuickAction quickAction) {
+		super(quickAction);
+	}
+
+	@Override
+	public void execute(final MapActivity activity) {
+		final GpxSelectionHelper selectedGpxHelper = activity.getMyApplication()
+			.getSelectedGpxHelper();
+		if (selectedGpxHelper.isShowingAnyGpxFiles()) {
+			selectedGpxHelper.clearAllGpxFileToShow();
+		} else {
+			selectedGpxHelper.restoreSelectedGpxFiles();
+
+			final File dir = activity.getMyApplication().getAppPath(IndexConstants.GPX_INDEX_DIR);
+			CallbackWithObject<GPXFile[]> callbackWithObject = new CallbackWithObject<GPXFile[]>() {
+				@Override
+				public boolean processResult(GPXFile[] result) {
+					return true;
+				}
+			};
+			selectedGpxHelper.restoreSelectedGpxFiles();
+			List<GpxSelectionHelper.SelectedGpxFile> selectedGpxFiles = selectedGpxHelper
+				.getSelectedGPXFiles();
+			List<String> files = new ArrayList<>();
+			for (GpxSelectionHelper.SelectedGpxFile file : selectedGpxFiles) {
+				files.add(file.getGpxFile().path.substring(file.getGpxFile().path.lastIndexOf("/")));
+			}
+
+			GpxUiHelper.loadFileInDifferentThread(activity, callbackWithObject, dir,
```

Comment:
you might already have these files read (selectedGpxFilesBackUp). why read them again every time when you call quick action?
1 - what if file was deleted? 
2 - what if app was restarted?

Reply:
Yes, overthinking. It works great without reloading.